### PR TITLE
Feat: onnx export cli

### LINKED ISSUE #3146: Feature: ONNX / TensorRT export for trained policies
### Ticket Type

Feature Request / Improvement

### Description

LeRobot covers the full pipeline from data collection to training to async inference, but there's currently no built-in way to export a trained policy (ACT, Diffusion Policy, etc.) to optimized formats like ONNX or TensorRT for edge deployment.

This is a common pain point for anyone deploying on resource-constrained hardware (Jetson Orin Nano, etc.) where PyTorch inference is too slow or memory-heavy. Right now the only option is D-Robotics' hardware-specific [rdk_LeRobot_tools](https://github.com/D-Robotics/rdk_LeRobot_tools), but there's nothing general-purpose.

### Proposed scope

A `lerobot-export` CLI command (or equivalent API) that:

1. **Loads a trained policy checkpoint** (local path or Hub repo)
2. **Exports to ONNX** with correct dynamic axes for batch size / sequence length
3. **Optionally optimizes with TensorRT** (FP16/INT8) when available
4. **Validates correctness** by comparing PyTorch vs exported model outputs on sample inputs

Example usage:
```bash
lerobot-export \
  --policy.path=lerobot/act_pusht \
  --format=onnx \
  --output=act_pusht.onnx

# With TensorRT optimization
lerobot-export \
  --policy.path=lerobot/act_pusht \
  --format=tensorrt \
  --precision=fp16 \
  --output=act_pusht.engine
```

### Why this matters

- The SO-101 + Jetson Orin Nano combo is increasingly popular and PyTorch inference can be a bottleneck
- ONNX export enables deployment across diverse hardware (Jetson, Intel NPUs, Qualcomm, etc.) without vendor lock-in
- TensorRT typically gives 2-5x inference speedup over vanilla PyTorch on NVIDIA GPUs
- This would complement the existing async inference stack by making the policy server itself faster

### What I've looked into

- `torch.onnx.export` works for ACT and Diffusion Policy architectures with some attention to dynamic axes
- The main challenge is handling policy-specific preprocessing/postprocessing (action chunking, denoising loops for diffusion, etc.)
- VLA models (Pi0, SmolVLA) are harder due to size but ONNX export is still feasible; TensorRT would require more work

Happy to put together a PR if this is something the team would want upstream. Would also appreciate guidance on which policies to prioritize.

### PR DESCRIPTION
**Pull request by tsuu-abj**
## Title

feat(export): add `lerobot-export` CLI for ONNX/TensorRT edge deployment

## Summary / Motivation

Adds a `lerobot-export` CLI that exports trained LeRobot policies to ONNX (and
optionally TensorRT) for edge deployment on devices such as Jetson Orin Nano/NX.
Users running trained policies on Jetson-class devices need an optimized
inference format; ONNX + TensorRT FP16 provides significant speedup over
PyTorch on SM_87 GPUs without requiring the Python training stack at inference
time.

## Related issues

- Fixes / Closes: #3146

## What changed

- **New CLI** `lerobot-export` (entry point `lerobot.scripts.lerobot_export`)
  supporting `--format=onnx` (default) and `--format=tensorrt`.
- **New `lerobot.export` package** with:
  - `ExportSpec` dataclass + `register_export_wrapper` plugin registry with
    auto-discovery by naming convention (`policies/<type>/export_<type>.py`).
  - Reusable adapter primitives: `DictBatchAdapter` (Pattern A) and
    `IterativeDenoisingAdapter` ABC (Pattern B) in `export/adapters/`.
  - Three ONNX exporter backends: `dynamo`, `legacy`, and `auto` (tries dynamo,
    falls back to legacy).
  - `validate_onnx()` — post-export parity check (max_abs_error, cos_sim,
    `torch.allclose`). Supports `--validation-trials=N` for random-input trials.
  - `export_to_tensorrt()` via `trtexec` subprocess with engine cache and
    hardware guards (FP8/INT8 on SM_87 raises; FP16 on SM<80 warns).
  - `save_normalization_stats()` + `NormalizedWrapper` (opt-in
    `--fold-normalization` bakes stats as ONNX constants).
- **Per-policy adapters** co-located with each policy module:
  - `policies/act/export_act.py` — ACT via `DictBatchAdapter`
  - `policies/diffusion/export_diffusion.py` — UNet-only and full DDIM loop
    (`--diffusion-mode=ddim-N`)
  - `policies/vqbet/export_vqbet.py` — VQ-BeT via `DictBatchAdapter`
- Unsupported policies (SAC, TDMPC, PI0, SmoLVLA) raise `NotImplementedError`
  with a concrete extension guide instead of silently producing wrong output.
- **New optional dependency group** `export` (`onnx`, `onnxruntime`,
  `onnxscript`) in `pyproject.toml`; `_onnx_available` / `_onnxruntime_available`
  added to `import_utils.py`.

No breaking changes to existing APIs.

## How was this tested (or how to run locally)

**Unit tests** (31 passed, 1 skipped on macOS/MPS):

```bash
uv sync --locked --extra export --extra test
uv run pytest tests/test_export.py tests/policies/test_export_act.py \
    tests/policies/test_export_vqbet.py tests/policies/test_export_diffusion.py -sv
```

**End-to-end validation against real Hub checkpoints** (macOS/CPU, opset 18,
`--validation-trials=5`):

```bash
# ACT
uv run lerobot-export \
  --policy.path=lerobot/act_aloha_sim_insertion_human \
  --output_path=./outputs/verify_act --device=cpu --exporter=legacy \
  --validation_trials=5
# => worst max_abs_error=2.09e-06, min cos_sim=1.000000, allclose=True

# Diffusion UNet (batch_size=4, dynamic)
uv run lerobot-export \
  --policy.path=lerobot/diffusion_pusht \
  --output_path=./outputs/verify_diffusion --device=cpu \
  --validation_trials=5
# => worst max_abs_error=9.06e-06, min cos_sim=1.000000, allclose=True
```

TensorRT engine builds require CUDA + `trtexec`; the hardware-guard error paths
are covered by `tests/test_export.py::test_export_to_tensorrt_raises_without_cuda`.

## Checklist (required before merge)

- [x] Linting/formatting run (`pre-commit run -a`)
- [x] All tests pass locally (`pytest`)
- [x] Documentation updated
- [ ] CI is green
- [ ] Community Review: I have reviewed another contributor's open PR and linked it here: # (insert PR number/link)

## Reviewer notes

- **ACT + dynamo batch>1 is a known limitation**: even `--exporter=dynamo`
  produces a fixed `batch_size=1` ONNX because `ACT.forward` allocates
  `torch.zeros([batch_size, latent_dim])` inside the model, causing
  `torch.export` to specialize on the concrete value. Fixing this requires
  modifying `ACT.forward` itself and is out of scope for this PR.
- **`export/adapters/`** contains the two reusable primitives intended to keep
  future per-policy adapters thin. Feedback on the API surface is welcome.
- **TensorRT path** is tested only via error-path unit tests (no CUDA on the
  development host); a GPU CI job is a follow-up.
- Older Hub checkpoints (e.g. `lerobot/act_aloha_sim_insertion_human`) do not
  ship `policy_preprocessor.json`, so `normalization_stats.json` is written as
  `{}` with a warning — this is expected and logged clearly.

### CODE REVIEW

**Comment by rylinjames** [comment]
hey, just flagging that VLA ONNX/TRT export for pi0, pi0.5, smolvla, and gr00t is already working in reflex-vla (github.com/FastCrest/reflex-vla) if it's useful reference for the PR.

the VLA export is the hard part. smolvla in particular took a bunch of patches to get right (broken kv cache wiring, wrong sinusoidal embedding, 5D vs 4D image dims from the AutoProcessor). we also bake the denoising loop unrolled into the monolithic ONNX so the TRT engine is a single graph. might save some time vs figuring it out from scratch.

**Comment by starsleep** [comment]
Thank you for submitting this PR. We are also very interested in ONNX/TensorRT conversion of LeRobot policies and deployment to edge devices.

Based on the latest commit of this PR, `dff23d90412b3bc8367cdc5a242b9ecc122de832`, we tested ACT and Diffusion Policy using the CUDA backend.

We hope the following benchmark results and issues we encountered are helpful for reviewing and improving the PR.

## Test Environment

| Item | Value |
| --- | --- |
| OS                     | Windows 11                                |
| GPU                    | NVIDIA GeForce RTX 4060 Laptop GPU (8 GB) |
| GPU Compute Capability | 8.9                                       |
| NVIDIA Driver          | 560.94                                    |
| Python                 | 3.12.13                                   |
| PyTorch                | 2.10.0+cu128                              |
| CUDA                   | 12.8                                      |
| cuDNN                  | 9.10.2                                    |
| ONNX Runtime GPU       | 1.26.0                                    |
| TensorRT               | 10.16.1.11                                |
| Warm-up Iterations     | 20                                        |
| Benchmark Iterations   | 100                                       |

All latency measurements were collected in the same test environment. For each runtime, we performed 20 warm-up iterations followed by 100 inference iterations for benchmarking.

---

## ACT Results

### Performance

Latency is reported in `ms`.

| Runtime | Precision | Mean | Std | P50 | P95 | P99 |
| --- | :---: | ---: | ---: | ---: | ---: | ---: |
| PyTorch                          | FP32 | 27.983    | 4.536 | 27.451    | 28.484    | 30.706    |
| ONNX Runtime                     | FP32 | 20.360    | 0.156 | 20.350    | 20.595    | 20.740    |
| TensorRT                         | FP32 | 21.431    | 1.011 | 21.771    | 22.761    | 22.976    |
| PyTorch                          | FP16 | 17.053    | 1.685 | 16.463    | 20.261    | 23.218    |
| ONNX Runtime                     | FP16 | 13.262    | 0.499 | 13.189    | 13.987    | 14.549    |
| TensorRT                         | FP16 | **6.094** | 1.094 | **5.728** | **9.417** | **9.455** |

In our environment, ONNX Runtime showed lower average latency than PyTorch for ACT FP32, while TensorRT FP32 also provided a similar level of improvement.

The performance improvement was more significant in FP16. The average latency decreased from `17.053 ms` with PyTorch FP16 to `6.094 ms` with TensorRT FP16.

### Accuracy

The outputs of each backend were compared against PyTorch using the same precision.

| Runtime | Precision | Output | Mean Abs Error | Max Abs Error | Cosine Similarity | Cosine Error |
| --- | :---: | --- | ---: | ---: | ---: | ---: |
| ONNX Runtime                                                                   | FP32 | `action_chunk` | 3.22791e-4 | 1.02255e-3 | 1.000000 | 3.57628e-7 |
| TensorRT                                                                       | FP32 | `action_chunk` | 4.56617e-5 | 3.08678e-4 | 1.000000 | 1.19209e-7 |
| ONNX Runtime                                                                   | FP16 | `action_chunk` | 1.60021e-3 | 8.11768e-3 | 0.999987 | 1.29938e-5 |
| TensorRT                                                                       | FP16 | `action_chunk` | 1.94020e-3 | 8.30078e-3 | 0.999994 | 6.49691e-6 |

For FP32, both ONNX Runtime and TensorRT showed very high numerical agreement with the PyTorch output.

For FP16, the absolute error increased compared with FP32, but cosine similarity remained very high.

---

## ACT FP16 dtype issue

**File:** `src/lerobot/policies/act/modeling_act.py`
**Lines:** `456-458`

While testing FP16 inference with `model.half()`, we encountered a dtype mismatch.

In the current inference path, the latent tensor is explicitly created as FP32:

```python
latent_sample = torch.zeros(
    [batch_size, self.config.latent_dim],
    dtype=torch.float32,
).to(batch[OBS_STATE].device)

```

When the entire model is converted to FP16, this can lead to a dtype mismatch between the FP32 latent tensor and subsequent FP16 layers.

For testing purposes, we modified the code so that the latent tensor follows the dtype of the input state:

```python
latent_sample = torch.zeros(
    [batch_size, self.config.latent_dim],
    dtype=batch[OBS_STATE].dtype,
).to(batch[OBS_STATE].device)

```

After this change, we were able to run ACT with PyTorch FP16 successfully.

It may therefore be worth considering whether this tensor should follow the input or model dtype instead of being explicitly fixed to `torch.float32`.

---

## Diffusion Policy Results

### Performance

Latency is reported in `ms`.

| Runtime | Precision | Mean | Std | P50 | P95 | P99 |
| --- | :---: | ---: | ---: | ---: | ---: | ---: |
| PyTorch                          | FP32 | 15.418    | 0.436     | 15.311    | 16.188    | 16.649    |
| ONNX Runtime                     | FP32 | 15.786    | 0.304     | 15.778    | 16.224    | 16.710    |
| TensorRT                         | FP32 | **5.988** | **0.063** | **5.978** | **6.097** | **6.146** |

For Diffusion Policy FP32, PyTorch and ONNX Runtime showed similar latency in our environment.

TensorRT, on the other hand, reduced the average latency to `5.988 ms`, compared with `15.418 ms` for PyTorch FP32.

TensorRT also showed relatively low latency variance in this test.

### Accuracy

| Runtime | Precision | Output | Mean Abs Error | Max Abs Error | Cosine Similarity | Cosine Error |
| --- | :---: | --- | ---: | ---: | ---: | ---: |
| ONNX Runtime                                                                   | FP32 | `denoised` | 1.16753e-4 | 3.31178e-4 | 1.000000 | 5.96046e-8 |
| TensorRT                                                                       | FP32 | `denoised` | 1.91200e-4 | 7.09891e-4 | 1.000000 | 5.96046e-8 |

For Diffusion Policy FP32, both backends also showed very high agreement with the PyTorch output.

---

## Diffusion Policy FP16 dtype issue

During FP16 testing of Diffusion Policy, we encountered dtype consistency issues in several places.

### Sinusoidal positional embedding

**File:** `src/lerobot/policies/diffusion/modeling_diffusion.py`
**Lines:** `589-596`

The positional embedding is currently generated as follows:

```python
half_dim = self.dim // 2
emb = math.log(10000) / (half_dim - 1)
emb = torch.exp(torch.arange(half_dim, device=device) * -emb)
emb = x.unsqueeze(-1) * emb.unsqueeze(0)
emb = torch.cat((emb.sin(), emb.cos()), dim=-1)

```

Because the dtype is not explicitly specified for `torch.arange()`, the intermediate embedding may be created as FP32 even when the model and input are using FP16.

This can lead to dtype mismatches when the embedding is later combined with FP16 tensors or passed through FP16 layers.

### Timestep embedding

**File:** `src/lerobot/policies/diffusion/modeling_diffusion.py`
**Lines:** `711-712`

We also observed a potential dtype mismatch between the timestep embedding and FP16 layers around the following code:

```python
timesteps_embed = self.diffusion_step_encoder(timestep)
timesteps_embed = timesteps_embed.to(dtype=x.dtype)

```

During testing, we tried explicitly converting some of the embeddings to `x.dtype`.

However, Diffusion Policy combines several intermediate tensors, including positional embeddings, timestep embeddings, and conditioning tensors. Because of this, it appears that dtype consistency may need to be handled across the full inference path rather than by converting only individual tensors.

The FP16 test currently fails with the following error:

```text
mat1 and mat2 must have the same dtype, but got Float and Half

```

For this reason, Diffusion Policy FP16 results were not included in this benchmark.

---

## `packaging.version` import

**File:** `src/lerobot/policies/pretrained.py`
**Lines:** `24-26`

The file directly uses `packaging.version.parse()`. In our test environment, we explicitly imported `packaging.version`:

```python
import packaging
import packaging.version
import safetensors

```

The relevant usage is located here:

**File:** `src/lerobot/policies/pretrained.py`
**Lines:** `141-143`

```python
if packaging.version.parse(safetensors.__version__) >= packaging.version.parse("0.4.3"):
    kwargs["device"] = map_location

```

Depending on the environment, `packaging.version` may not always be implicitly available through the top-level `packaging` module, so an explicit import may be slightly safer.

---

## TensorRT 10.x / 11.x compatibility

**File:** `src/lerobot/export/tensorrt_export.py`
**Lines:** `132-136`

We also observed a difference in the accepted `--memPoolSize` argument format depending on the TensorRT / `trtexec` version.

The current code uses:

```python
f"--memPoolSize=workspace:{workspace_gb * 1024}MiB",

```

In our test environment with TensorRT `10.16.1.11`, we used the following form instead:

```python
f"--memPoolSize=workspace:{workspace_gb * 1024}",

```

There may be differences between TensorRT 10.x and 11.x not only in the Python API, but also in `trtexec` CLI behavior and dependencies. It may therefore be worth checking this part against the range of TensorRT versions that the project intends to support.

This benchmark was conducted using TensorRT `10.16.1.11`.

---

## Edge deployment / custom inference engine

Finally, we are developing a **C++-based, cross-GPU-vendor inference engine for edge devices**.

The engine is designed to run without a Python runtime, load models directly through our own Safetensors parser, and provide a common inference pipeline across NVIDIA GPUs as well as GPUs and edge platforms from other vendors.

In the longer term, we are also exploring the possibility of loading and executing LeRobot policies directly in our engine, rather than relying exclusively on specific runtimes such as ONNX Runtime or TensorRT.

We are particularly interested in the following use cases:

- Lightweight C++ inference runtime without Python dependencies
- Cross-GPU-vendor backend support
- Direct model loading from Safetensors
- Low-latency inference on edge devices
- Portable deployment of LeRobot policies

We would be very interested to know whether the LeRobot project would also be interested in this kind of **vendor-neutral edge deployment** or **custom C++ inference engine integration**.

Thank you again for the PR and for the work on this area.
